homeassistant/components/zha/__init__.py

```diff
@@ -0,0 +1,293 @@
+
+    @asyncio.coroutine
+    def async_device_initialized(self, device, join):
+        """Handle device joined and basic information discovered (async)."""
+        import bellows.zigbee.profiles
+        import homeassistant.components.zha.const as zha_const
+
+        for endpoint_id, endpoint in device.endpoints.items():
+            if endpoint_id == 0:  # ZDO
+                continue
+
+            discovered_info = yield from _discover_endpoint_info(endpoint)
+
+            has_valid_profile = False
+            component = None
+            used_clusters = []
+            device_key = '%s-%s' % (str(device.ieee), endpoint_id)
+            node_config = self._config[DOMAIN][CONF_DEVICE_CONFIG].get(
+                device_key, {})
+
+            if endpoint.profile_id in bellows.zigbee.profiles.PROFILES:
+                profile = bellows.zigbee.profiles.PROFILES[endpoint.profile_id]
+                if zha_const.DEVICE_CLASS.get(endpoint.profile_id,
+                                              {}).get(endpoint.device_type,
+                                                      None):
+                    used_clusters = profile.CLUSTERS[endpoint.device_type]
+                    has_valid_profile = True
+                    profile_info = zha_const.DEVICE_CLASS[endpoint.profile_id]
+                    component = profile_info[endpoint.device_type]
+
+            if DEVICE_TYPES.get(endpoint.profile_id,
+                                {}).get(endpoint.device_type, None):
+                has_valid_profile = True
+                info = DEVICE_TYPES[endpoint.profile_id][endpoint.device_type]
+                component, used_clusters = info
+
+            if ha_const.CONF_TYPE in node_config:
+                component = node_config[ha_const.CONF_TYPE]
+
+            if has_valid_profile and component:
+                clusters = [endpoint.clusters[c] for c in used_clusters if c in
+                            endpoint.clusters]
+                discovery_info = {
+                    'endpoint': endpoint,
+                    'clusters': clusters,
+                    'new_join': join,
+                }
+                discovery_info.update(discovered_info)
+
+                yield from discovery.async_load_platform(
+                    self._hass,
+                    component,
+                    DOMAIN,
+                    discovery_info,
+                    self._config,
+                )
+
+            for cluster_id, cluster in endpoint.clusters.items():
+                cluster_type = type(cluster)
+                if cluster_id in used_clusters:
+                    continue
+                if cluster_type not in zha_const.SINGLE_CLUSTER_DEVICE_CLASS:
+                    continue
+
+                component = zha_const.SINGLE_CLUSTER_DEVICE_CLASS[cluster_type]
+                discovery_info = {
+                    'endpoint': endpoint,
+                    'clusters': [cluster],
```

Comment:
Someone ran into this while testing, so good catch.

I don't grok everything you're saying - I'll experiment and get a better understanding, then see what the best option is. Thanks for the pointers.